### ISSUE TITLE
Using an optional call with `.flat()` produces incorrect types

### ISSUE THREAD

**Issue by tom-sherman** [microsoft/TypeScript#41007]
**TypeScript Version:**  4.0.2

**Search Terms:**

flat, flatMap, undefined, array, optional chaining, optional calling

**Code**

```ts repro
// @target: es2019

declare const arr: undefined | number[];
const x = arr?.flat() // (number | undefined)[] | undefined
//    ^?
```

**Expected behavior:**

`x` should have the type `undefined | number[]`

**Actual behavior:**

`x` has the type `(number | undefined)[] | undefined`.

In general, calling `?.flat(depth: D)` on `NestedArray<T>` seems to result in the type of `FlatArray<T | undefined, D>  | undefined` when it should have the type `FlatArray<T, D> | undefined`

**Playground Link:** https://www.typescriptlang.org/play?target=99&ts=4.0.2#code/CYUwxgNghgTiAEYD2A7AzgF3rGAueAriqAGYCWKIw8APvCgQLYBGIMA2gLoDcAUL8nRYAHvAC82GDAD8AOhLQMACgCU8APTr4Shiza1CxEOUrAVXA0VIUqQA

**Related Issues:**

**Comment by WorldSEnder** [comment]
Hand desugaring to

```
const x = (arr === null || arr === undefined) ? undefined : arr.flat()
```

works as expected.

**Comment by tom-sherman** [comment]
This seems to have been fixed in 4.2.x

**Comment by rbuckton** [comment]
I can confirm this has been fixed for some time now.
